### PR TITLE
cholesky_solve_backward: speed up using output_mask

### PR DESCRIPTION
**Pull request by mlubin**
Introduces a faster path for `cholesky_solve_backward` when the gradient with respect to the cholesky factor isn't required.

Adds test coverage in `test_linalg.py`.

# Example

## Setup

```py
import torch
torch.set_num_threads(1)
mat = torch.randn(500, 1000)
mat = mat @ mat.T
L = torch.linalg.cholesky(mat, upper=False)

rhs = torch.randn(500, 1)
rhs.requires_grad = True

sol = torch.cholesky_solve(rhs, L, upper=False).sum(dim=0)
```

## Before
```
%timeit torch.autograd.grad(sol, rhs, retain_graph=True)
2.61 ms ± 18.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

## After
```
%timeit torch.autograd.grad(sol, rhs, retain_graph=True)
109 µs ± 3.42 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

### CODE REVIEW

**Review by albanD — COMMENTED** [review]
The change sounds good thanks!
@lezcano where are the autograd tests for this?

**Review by lezcano — COMMENTED** [review]
There is an op-info at `common_methods_invocations.py`, but that will not test this new route. I would suggest you add a one-off test in `test_linalg.py`

File: test/test_linalg.py [hunk 1]
```diff
@@ -2637,6 +2637,27 @@ def test_cholesky_solve_out_errors_and_warnings(self, device, dtype):
...
+            upper = False
+
+            b = torch.randn(*b_dims, dtype=dtype, device=device, requires_grad=True)
+            A = random_hermitian_pd_matrix(*A_dims, dtype=dtype, device=device)
+            A.requires_grad = requires_A_grad
+            L = torch.linalg.cholesky(A, upper=upper)
+            (torch.cholesky_solve(b, L, upper=upper).sum()).backward()
+
+            expected_grad = torch.linalg.solve(A, torch.ones_like(b))
+            self.assertEqual(b.grad, expected_grad)
```
Comment (lezcano) [inline test/test_linalg.py:2659]: Rather than testing this way, which assumes which are the formulas for the gradinet, could you use [gradcheck](https://pytorch.org/docs/stable/generated/torch.autograd.gradcheck.html)?

Comment (mlubin) [reply test/test_linalg.py]: Updated the test to use `gradcheck`. I wasn't able to find a clean way to use gradcheck for the matrix part though. `cholesky_solve` is a bit weird because it ignores the lower or upper triangle of the matrix and computes the gradient as if the matrix is symmetric. When finite differencing, you just see zeros in the components that are ignored.

File: test/test_linalg.py [hunk 2]
```diff
@@ -2637,6 +2637,20 @@ def test_cholesky_solve_out_errors_and_warnings(self, device, dtype):
...
+    def test_cholesky_solve_backward(self, device, dtype):
+        b_dims = (5, 2)
+        L_dims = (5, 5)
+        upper = False
+
+        b = torch.randn(*b_dims, dtype=dtype, device=device, requires_grad=True)
+        L = torch.randn(*L_dims, dtype=dtype, device=device)
+        torch.autograd.gradcheck(lambda b: torch.cholesky_solve(b, L, upper=upper), (b,))
+        # TODO: Also check the gradient with respect to L. This seems to be messy with
+        # gradcheck because gradcheck isn't aware of the implicit symmetry in L.
```
Comment (lezcano) [inline test/test_linalg.py:2652]: For this to work you would need to create an aux function that takes in unconstrained matrices, preprocesses them so that they are valid inputs for `cholesky_solve` (i.e. it calls `tril() or `triu()`) and then calls `cholesky_solve`.

**Review by lezcano — APPROVED** [review]

File: test/test_linalg.py (hunk 1 above)
Comment (lezcano) [reply test/test_linalg.py]: left a comment below

File: test/test_linalg.py (hunk 2 above)
Comment (mlubin) [reply test/test_linalg.py]: An extra `tril()` seems to fix the issue, thanks!

**Comment by lezcano** [comment]
Yep, I also thought about that nit with `upper=False`. Thank you for fixing it!
